### PR TITLE
PERF: Dematerialize topic_reply_count

### PR DESCRIPTION
**Pull request by riking**
Instead of updating topic_reply_count on every post creation, deletion, and ownership change, calculate it from scratch during trust level promotions. Additionally, replace the `topic_count + reply_count` on the client with `any_posts`, as it was only ever compared to 0.

### CODE REVIEW

File: app/models/user_stat.rb [hunk 1]
```diff
@@ -151,12 +151,24 @@ def update_distinct_badge_count
...
 
   # topic_reply_count is a count of posts in other users' topics
-  def update_topic_reply_count
-    self.topic_reply_count = Topic
-      .joins("INNER JOIN posts ON topics.id = posts.topic_id AND topics.user_id <> posts.user_id")
-      .where("posts.deleted_at IS NULL AND posts.user_id = ?", self.user_id)
-      .distinct
-      .count
+  def calc_topic_reply_count!(max)
+    self.topic_reply_count = DB.query_single(<<~SQL, self.user_id, max).first
```
Comment (riking) [inline app/models/user_stat.rb:155]: the column is now useless, drop it from the db

File: app/assets/javascripts/discourse/app/components/notification-consent-banner.js [hunk 2]
```diff
@@ -20,21 +20,19 @@ export default DesktopNotificationConfig.extend({
...
-    "currentUser.topic_count"
+    "currentUser.any_posts"
   )
   showNotificationPromptBanner(
     isNotSupported,
     isEnabled,
     bannerDismissed,
-    replyCount,
-    topicCount
+    any_posts
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/notification-consent-banner.js:29]: We use `camelCase` on the client-side (except very few places) 😉 

```suggestion
    anyPosts
```

File: app/models/trust_level3_requirements.rb [hunk 3]
```diff
@@ -143,11 +143,7 @@ def min_days_visited
   end
 
   def num_topics_replied_to
-    @user.posts
-      .public_posts
-      .where("posts.created_at > ? AND posts.post_number > 1", time_period.days.ago)
-      .select("distinct topic_id")
-      .count
+    @user.user_stat.calc_topic_reply_count!(min_topics_replied_to * 2, time_period.days.ago)
```
Comment (ZogStriP) [inline app/models/trust_level3_requirements.rb:146]: Why `*2` and not just `+1`?

File: app/models/user_stat.rb [hunk 4]
```diff
@@ -151,12 +154,30 @@ def update_distinct_badge_count
...
+    sql = <<~SQL
+    SELECT COUNT(*) count
+    FROM (
+      SELECT DISTINCT posts.topic_id
+      FROM posts
+      INNER JOIN topics ON topics.id = posts.topic_id
+      WHERE posts.user_id = ?
+      AND topics.user_id <> posts.user_id
+      AND posts.deleted_at IS NULL AND topics.deleted_at IS NULL
+      AND topics.archetype <> 'private_message'
```
Comment (ZogStriP) [inline app/models/user_stat.rb:167]: Looks like this is a new condition. Is it intended?

File: app/assets/javascripts/admin/models/tl3-requirements.js [hunk 5]
```diff
@@ -12,6 +12,11 @@ export default EmberObject.extend({
     return Math.round((minDaysVisited * 100) / timePeriod);
   },
 
+  @discourseComputed("num_topics_replied_to", "min_topics_replied_to")
+  capped_topics_replied_to(numReplied, minReplied) {
+    return numReplied === 2 * minReplied;
```
Comment (ZogStriP) [inline app/assets/javascripts/admin/models/tl3-requirements.js:17]: I think `>=` would be more correct here.

```suggestion
    return numReplied >= minReplied;
```

Comment (riking) [reply app/assets/javascripts/admin/models/tl3-requirements.js:17]: If I'm doing a +1, then `>` would be correct, no?

File: app/models/trust_level3_requirements.rb (hunk 3 above)
Comment (riking) [reply app/models/trust_level3_requirements.rb:146]: TL3 requirements checked for `.public_posts` https://github.com/discourse/discourse/pull/9769/files#diff-71b275fc6f97d344de4978707d0f2fdeL147

File: app/models/user_stat.rb (hunk 4 above)
Comment (riking) [reply app/models/user_stat.rb:167]: TL3 requirements checked for `.public_posts` https://github.com/discourse/discourse/pull/9769/files#diff-71b275fc6f97d344de4978707d0f2fdeL147

File: app/assets/javascripts/admin/models/tl3-requirements.js (hunk 5 above)
Comment (ZogStriP) [reply app/assets/javascripts/admin/models/tl3-requirements.js:17]: I'm always off by one :P

File: app/models/user_stat.rb (hunk 4 above)
Comment (ZogStriP) [reply app/models/user_stat.rb:167]: TL2 didn't as far as I know, but I guess it's better that way 👍

**Pull request merged**